libsignal-service/src/provisioning/manager.rs

```diff
@@ -235,48 +171,145 @@ impl<P: PushService> LinkingManager<P> {
+                        fetches_messages: bool,
+                        name: String,
+                        registration_id: u32,
+                        pni_registration_id: u32,
+                        capabilities: Capabilities,
+                    }
+
+                    #[derive(Debug, Serialize)]
+                    #[serde(rename_all = "camelCase")]
+                    struct Capabilities {
+                        pni: bool,
+                    }
+
+                    let encrypted_device_name = base64::encode(
+                        encrypt_device_name(
+                            csprng,
+                            device_name,
+                            &aci_public_key,
+                        )?
+                        .encode_to_vec(),
+                    );
+
+                    let profile_key = ProfileKey::create(
+                        profile_key
+                            .as_slice()
+                            .try_into()
+                            .map_err(ProvisioningError::InvalidProfileKey)?,
+                    );
+
+                    let request = LinkRequest {
+                        verification_code: provisioning_code,
+                        account_attributes: AccountAttributes {
                             registration_id,
                             pni_registration_id,
-                            service_ids: ServiceIds {
-                                aci: aci_uuid,
-                                pni: pni_uuid,
-                            },
-                            aci_private_key,
-                            aci_public_key,
-                            pni_private_key,
-                            pni_public_key,
-                            profile_key,
+                            fetches_messages: true,
+                            capabilities: Capabilities { pni: true },
+                            name: encrypted_device_name,
                         },
+                        aci_signed_pre_key: aci_signed_pre_key.try_into()?,
+                        pni_signed_pre_key: pni_signed_pre_key.try_into()?,
+                        aci_pq_last_resort_pre_key: aci_pq_last_resort_pre_key
+                            .try_into()?,
+                        pni_pq_last_resort_pre_key: pni_pq_last_resort_pre_key
+                            .try_into()?,
+                    };
+
+                    #[derive(Debug, Deserialize)]
+                    #[serde(rename_all = "camelCase")]
+                    struct LinkResponse {
+                        #[serde(rename = "uuid")]
+                        aci: Uuid,
+                        pni: Uuid,
+                        device_id: u32,
+                    }
+
+                    let LinkResponse {
+                        aci,
+                        pni,
+                        device_id,
+                    } = self
+                        .push_service
+                        .put_json(
```

Comment:
I think this should become a method of the push service. Then the embedded structs can be also moved there.

Reply:
Done in 61113e